homeassistant/components/keyboard_remote/__init__.py

```diff
@@ -36,177 +42,256 @@
+                    del self.active_handlers_by_descriptor[descriptor]
+                    await handler.async_stop_monitoring()
+                elif (
+                    (event.flags & aionotify.Flags.CREATE)
+                    or (event.flags & aionotify.Flags.ATTRIB)
+                ) and not descriptor_active:
+                    dev, handler = self.get_device_handler(descriptor)
+                    if handler is None:
+                        continue
+                    self.active_handlers_by_descriptor[descriptor] = handler
+                    await handler.async_start_monitoring(dev)
+        except asyncio.CancelledError:
+            return
+
+    class DeviceHandler:
+        """Manage input events using evdev with asyncio."""
+
+        def __init__(self, hass, dev_block):
+            """Fill configuration data."""
 
-            if device_descriptor is not None or device_name is not None:
-                thread = KeyboardRemoteThread(
-                    hass, device_name, device_descriptor, key_value
+            self.hass = hass
+
+            key_types = dev_block.get(TYPE)
+
+            self.key_values = set()
+            for key_type in key_types:
+                self.key_values.add(KEY_VALUE[key_type])
+
+            self.emulate_key_hold = dev_block.get(EMULATE_KEY_HOLD)
+            self.emulate_key_hold_delay = dev_block.get(EMULATE_KEY_HOLD_DELAY)
+            self.emulate_key_hold_repeat = dev_block.get(EMULATE_KEY_HOLD_REPEAT)
+            self.monitor_task = None
+            self.dev = None
+
+        async def async_keyrepeat(self, path, name, code, delay, repeat):
+            """Emulate keyboard delay/repeat behaviour by sending key events on a timer."""
+
+            await asyncio.sleep(delay)
+            while True:
+                self.hass.bus.async_fire(
+                    KEYBOARD_REMOTE_COMMAND_RECEIVED,
+                    {KEY_CODE: code, DEVICE_DESCRIPTOR: path, DEVICE_NAME: name},
+                )
+                await asyncio.sleep(repeat)
+
+        async def async_start_monitoring(self, dev):
+            """Start event monitoring task and issue event."""
+            if self.monitor_task is None:
+                self.dev = dev
+                self.monitor_task = self.hass.async_create_task(
+                    self.async_monitor_input(dev)
                 )
-                self.threads.append(thread)
+                self.hass.bus.async_fire(
+                    KEYBOARD_REMOTE_CONNECTED,
+                    {DEVICE_DESCRIPTOR: dev.path, DEVICE_NAME: dev.name},
+                )
+                _LOGGER.debug("Keyboard (re-)connected, %s", dev.name)
+
+        async def async_stop_monitoring(self):
+            """Stop event monitoring task and issue event."""
+            if self.monitor_task is not None:
+                try:
+                    self.dev.ungrab()
+                except (OSError, IOError):
```

Comment:
`IOError` is deprecated and has been replaced by `OSError`. We can remove `IOError`.